articles/quickstart/spa/auth0-react/01.md

```diff
@@ -0,0 +1,297 @@
+
+## Protecting a Route
+
+Your React application may implement navigation using a router. However, some routes may render public resources, while others may render private ones. In such a case, you can create an authentication route guard to render any component that defines a private resource conditionally. If Auth0 has authenticated the user, the guard renders the component; otherwise, the guard redirects the user to the [Auth0 Universal Login Page](https://auth0.com/docs/universal-login) to either log in or sign up.
+
+⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️
+
+**TODO: Final URL for additional routing examples.**
+ 
+⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️
+
+> React has no opinion on routing. This example assumes you are using [`react-router-dom`](https://github.com/ReactTraining/react-router#readme) &mdash; [access more examples for Gatsby and Next.js.](https://github.com/auth0-samples/auth0-react-sdk-sample)
+
+The default behavior is to redirect your users to the value of the `redirect_uri` prop of `Auth0Provider` after they log in or sign up using Auth0. However, you may want to redirect users directly to the private route they were trying to access. You can define such behavior using the `onRedirectCallback` prop of `Auth0Provider`.
+
+With React Router, your goal is to use the `history` object within the `onRedirectCallback` method to navigate to that private route programmatically. React Router exposes the `useHistory()` custom React Hook to make it easy to access the `history` object in any component that is a descendant of the `BrowserRouter` component.
+
+As such, you can create an `Auth0ProviderWithHistory` function component that renders `Auth0Provider`. This wrapper component then calls `useHistory()` within its body to make the `history` object available within the scope of `Auth0Provider` and its  `onRedirectCallback` event handler.
+
+**Root component:**
+
+```javascript
+ReactDOM.render(
+  <BrowserRouter>
+    <Auth0ProviderWithHistory>
+      <App />
+    </Auth0ProviderWithHistory>
+  </BrowserRouter>,
+  document.getElementById("root")
+);
+```
+
+**`Auth0ProviderWithHistory` component:**
+
+```javascript
+const Auth0ProviderWithHistory = ({ children }) => {
+  const history = useHistory();
+
+  const onRedirectCallback = (appState) => {
+    history.push(appState?.returnTo || window.location.pathname);
+  };
+
+  return (
+    <Auth0Provider
+      domain="${account.namespace}"
+      client_id="${account.clientId}"
+      redirect_uri={window.location.origin}
+      onRedirectCallback={onRedirectCallback}
```

Comment:
Why does `onRedirectCallback` use came case with client_id and redirect_uri do not? We need consistency here.

Reply:
I was wondering the same when I encountered it 🤔 It's also idiomatic to use camel case for variable names in JS/TS.

Reply:
To be fair this has come from the SPA SDK which has the same issue. I think the distinction is that `client_id` refers to a param on the query string when calling `/authorize`.

Not saying that's correct, that's just the way it is today. I would like to change this in the SPA SDK but will require a new major. We could address this now in the React SDK and map the params in the meantime.

Reply:
We can certainly rename the main one's (`client_id` etc). But while the SPA JS SDK let's you overload the constructor with any query params you want to send on the `/authorize` url - there will always be the possibility of mixing snake_case and camelCase.

We could do what the Express SDK does and use a container property like `authorizationParams` (https://github.com/auth0/express-openid-connect/blob/master/API.md#authorization-params-key) to pass extra stuff to `/authorize` I'll raise an issue to look into this.